### PR TITLE
improvement(test-cases): Add bare metal performance test

### PR DESCRIPTION
**Pull request by k0machi**
This commit fixes some issues regarding bare metal instances -
specifically it fixes setup for loader and monitor nodes, adds a new
keystore method to retrive bare metal cluster information instead of
storing it in the test case and fixes syslong-ng setup script to make
sure it reinstalls if it's present - to make repeated runs possible as
previously it would just do nothing and syslog-ng would transfer logs to
an old sct runner instead. Some similar fixes for prometheus and
scylla_setup scripts to make sure we can run them multiple times on the
same instance.

Task: scylladb/qa-tasks#1233

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by k0machi** [comment]
There are currently two major problems right now:

1. SCT is unable to cope with a node that doesn't have an nvme drive in a performance test - removing RAID setup code leads to later errors when tests try to execute `fstrim` on an arbitrary block device.
2. The runs are not repeatable - SCT tries to reinstall (update) scylla and fails when scylla_setup and other setup scripts return non-zero exit codes due to actions of that script being already done, e.g. "RAID already configured"

File: sdcm/tester.py
Comment (aleksbykov) [inline sdcm/tester.py:1367]: I think better to store ssh_username in config files. May add new parameters `baremetal_db_ssh_username`, `baremetal_loader_ssh_username` or similar. Because it will be also need in future the ssh key file, which should be stored in aws key store

**Comment by aleksbykov** [comment]
> There are currently two major problems right now:
> 
> 1. SCT is unable to cope with a node that doesn't have an nvme drive in a performance test - removing RAID setup code leads to later errors when tests try to execute `fstrim` on an arbitrary block device.

I think it should not failed without nvme drives. during setup if 1 disk or several but they are not nvme detected or passed scylla should be configure correctly.

> 2. The runs are not repeatable - SCT tries to reinstall (update) scylla and fails when scylla_setup and other setup scripts return non-zero exit codes due to actions of that script being already done, e.g. "RAID already configured"

i think we need to refactor setup for baremetal, and add checks that if hardware is already configured, don't run that scripts and only remove scylla configs and data dirs and reinstall it.

**Comment by roydahan** [comment]
> 



> > There are currently two major problems right now:
> > 
> > 1. SCT is unable to cope with a node that doesn't have an nvme drive in a performance test - removing RAID setup code leads to later errors when tests try to execute `fstrim` on an arbitrary block device.
> 
> I think it should not failed without nvme drives. during setup if 1 disk or several but they are not nvme detected or passed scylla should be configure correctly.

It's ok, we don't need to handle this if this is not handled at the moment.
We can assume that Scylla must have NVMe.
> 
> > 2. The runs are not repeatable - SCT tries to reinstall (update) scylla and fails when scylla_setup and other setup scripts return non-zero exit codes due to actions of that script being already done, e.g. "RAID already configured"
> 
> i think we need to refactor setup for baremetal, and add checks that if hardware is already configured, don't run that scripts and only remove scylla configs and data dirs and reinstall it.

Another option is to use "reuse_cluster" with "update_db_packages" in case we want to install new scylla.
> 
> do you have jobs or logs?

File: sdcm/cluster_baremetal.py
Comment (fruch) [inline sdcm/cluster_baremetal.py:22]: We need to be extra careful in changing baremetal implementation, since the siren tests are using it, so test it with siren at some point before merging this

File: test-cases/performance/perf-regression-throughput-baremetal-5gb.yaml
Comment (fruch) [inline test-cases/performance/perf-regression-throughput-baremetal-5gb.yaml:27]: We need to see how this part of the configuration isn't coming from public source code

File: sdcm/tester.py
Comment (fruch) [reply sdcm/tester.py:1367]: I don't understand what's wrong with the configuration option we have today for the db user, why can't we use that ?

Why have other defaults in the code, or why introduce new configuration options ?

**Comment by fruch** [comment]
> > 
> 
> 
> 
> > > There are currently two major problems right now:
> > > 
> > > 1. SCT is unable to cope with a node that doesn't have an nvme drive in a performance test - removing RAID setup code leads to later errors when tests try to execute `fstrim` on an arbitrary block device.
> > 
> > I think it should not failed without nvme drives. during setup if 1 disk or several but they are not nvme detected or passed scylla should be configure correctly.
> 
> It's ok, we don't need to handle this if this is not handled at the moment.
> We can assume that Scylla must have NVMe.
> > 
> > > 2. The runs are not repeatable - SCT tries to reinstall (update) scylla and fails when scylla_setup and other setup scripts return non-zero exit codes due to actions of that script being already done, e.g. "RAID already configured"
> > 
> > i think we need to refactor setup for baremetal, and add checks that if hardware is already configured, don't run that scripts and only remove scylla configs and data dirs and reinstall it.
> 
> Another option is to use "reuse_cluster" with "update_db_packages" in case we want to install new scylla.

update_db_packages might work 

We can also implement cleanup for bare metal, that would clear scylla installation and unmount disks, and what's needed to make install work again after that.

> > 
> > do you have jobs or logs?
> 
>

**Comment by k0machi** [comment]
> do you have jobs or logs?

I am currently starting the job manually via hydra, so no logs unfortunately, once I stabilize it enough to at least pass once I'll run it in jenkins

> I don't understand what's wrong with the configuration option we have today for the db user, why can't we use that ?

This is just a temporary plug to satisfy the class, as for `db_user`, is there also `loader_user` param? I haven't touched that aside that because I wasn't sure if I should be adding new params or not

> update_db_packages might work

Will try

> We need to see how this part of the configuration isn't coming from public source code

Can't we pass it as env as with other SCT Config values?

> We need to be extra careful in changing baremetal implementation, since the siren tests are using it, so test it with siren at some point before merging this

Hm, I was under impression this wasn't used much, can @enaydanov elaborate?

> I think it should not failed without nvme drives. during setup if 1 disk or several but they are not nvme detected or passed scylla should be configure correctly.

It's not a configuration issue, there are places later during stress that just run `fstrim` and if the disk doesn't support it, it errors out and causes a shutdown. It seems that EBS Virtual Block devices do not support that operation.

```
xvda    202:0    0    30G  0 disk
└─xvda1 202:1    0    30G  0 part /
nvme0n1 259:0    0 442.4G  0 disk /var/lib/scylla
scyllaadm@argusdb-node-1:~$ sudo fstrim /
fstrim: /: the discard operation is not supported
```

**Comment by k0machi** [comment]
Ready for review, more or less, will add a jenkins pipeline in a moment, need to shorten the normal parallel one to only contain one test that (maybe) can be specified in the params (so you could have one job that could do read, write and mixed). Or maybe just split it into 3 jobs.

File: sdcm/cluster_baremetal.py
Comment (k0machi) [reply sdcm/cluster_baremetal.py:22]: credentials.name here is ssh keyfile comment - it doesn't seem it would work with pretty much any key we have.

File: sdcm/cluster_baremetal.py
Comment (temichus) [inline sdcm/cluster_baremetal.py:68]: please leave a comment here that describe this if statement logic

File: sdcm/tester.py
Comment (fruch) [inline sdcm/tester.py:1358]: you gonna have multiple of those, and not just one set of configuration... 
probably need to name them and have a configuration option to point to them

similar to how we do it fro the internal-qa repo, as in the 5000 table case:

```yaml
cs_user_profiles:
    - scylla-qa-internal/cust_d/templated_tables_mv.yaml
```

File: sdcm/provision/common/utils.py
Comment (fruch) [inline sdcm/provision/common/utils.py:155]: why do we need this reinstall ? why isn't it enough we have syslog-ng installed ?

File: sdcm/cluster_baremetal.py
Comment (fruch) [inline sdcm/cluster_baremetal.py:179]: what's special about targets ? and why do we need to pop it out ?

File: test-cases/performance/perf-regression-throughput-baremetal-5gb.yaml
Comment (fruch) [inline test-cases/performance/perf-regression-throughput-baremetal-5gb.yaml:13]: I think this should be removed, we should have code using it beside provisioning code

File: sdcm/cluster_baremetal.py
Comment (fruch) [inline sdcm/cluster_baremetal.py:109]: I don't understand the need of this property, also we don't want default like `ubuntu` to be in the code.
this should have it's own new sct configuration option

File: sdcm/keystore.py
Comment (fruch) [inline sdcm/keystore.py:104]: this name is misleading, since it's more then just credentials.
it's also the information of a specific set of baremetal setup

File: sdcm/tester.py
Comment (k0machi) [reply sdcm/tester.py:1358]: Yes, I thought of that. I can implement this in this PR if needed.

File: test-cases/performance/perf-regression-throughput-baremetal-5gb.yaml
Comment (k0machi) [reply test-cases/performance/perf-regression-throughput-baremetal-5gb.yaml:13]: Will do

File: sdcm/cluster_baremetal.py
Comment (k0machi) [reply sdcm/cluster_baremetal.py:109]: I've moved this into bare metal config already, maybe just raise an exception there if it's unset?

File: sdcm/keystore.py
Comment (k0machi) [reply sdcm/keystore.py:104]: Will change

File: sdcm/cluster_baremetal.py
Comment (k0machi) [reply sdcm/cluster_baremetal.py:179]: PhysicalMachineCluster -> BaseCluster has no parameter called `targets` and so the constructor raises an exception.

File: sdcm/keystore.py
Comment (k0machi) [reply sdcm/keystore.py]: Done

File: sdcm/cluster_baremetal.py
Comment (k0machi) [reply sdcm/cluster_baremetal.py:109]: Changed to re-raise AttributeError

File: sdcm/tester.py
Comment (k0machi) [reply sdcm/tester.py]: Done, implemented as `s3_baremetal_config`

**Comment by k0machi** [comment]
Here's a [link](https://argus.scylladb.com/test/4224a5ba-77ce-4d0a-a4c3-c424a657db17/runs?additionalRuns[]=3e8eb05d-0612-4973-bedd-701a3f7a1043) to the test job in Argus.

**Review by roydahan — COMMENTED** [review]
Overall looks good to me.
As @fruch mentioned we need to make sure @enaydanov review it and possibly test siren longevity making sure it doesn't break anything.

Config and example looks convenient to me.

File: test-cases/performance/perf-regression-throughput-baremetal-5gb.yaml
Comment (k0machi) [reply test-cases/performance/perf-regression-throughput-baremetal-5gb.yaml:13]: There's a slight problem with changing this - in the pipeline the SCT Runner step actually runs with `AWS` configuration, to create a runner inside AWS, so I'm not sure if I can remove it (talking about `instance_type_db`)

**Comment by roydahan** [comment]
@k0machi is it waiting to you or reviewers?

**Comment by k0machi** [comment]
> @k0machi is it waiting to you or reviewers? 

Waiting for Evgeniy to review, he should do it tomorrow.

**Comment by enaydanov** [comment]
@k0machi can you rebase this PR to the latest master?

**Comment by k0machi** [comment]
> @k0machi can you rebase this PR to the latest master?

Done @enaydanov

**Comment by roydahan** [comment]
@enaydanov did you review this?

File: sdcm/cluster_baremetal.py
Comment (enaydanov) [reply sdcm/cluster_baremetal.py:22]: Can you change it to something like `getattr(self.parent_cluster, "ssh_username", credentials.name)` ?

File: sdcm/cluster_baremetal.py
Comment (enaydanov) [reply sdcm/cluster_baremetal.py:109]: Why do you need to catch exception at all in this case?

**Review by enaydanov — CHANGES_REQUESTED** [review]

**Comment by k0machi** [comment]
> Why do you need to catch exception at all in this case?
It's to shut up the pylint about `no-member`, just to be explicit that we may be accessing non-existent attribute there

> Can you change it to something like getattr(self.parent_cluster, "ssh_username", credentials.name) ?
Done

**Comment by enaydanov** [comment]
> > Why do you need to catch exception at all in this case?
> 
> It's to shut up the pylint about `no-member`, just to be explicit that we may be accessing non-existent attribute there
> 
In that case it's better to use explicit `# pylint: disable=...` or add a comment. It's not obvious.

Also, if you'll prefer to stay with try/except, it's better to use just `raise` to not add a duplicate to a traceback.

File: sdcm/cluster_baremetal.py
Comment (enaydanov) [inline sdcm/cluster_baremetal.py:137]: This is also a problem. Need to change it to `kwargs.pop('ssh_username', None)`

**Review by enaydanov — CHANGES_REQUESTED** [review]

File: sdcm/cluster_baremetal.py
Comment (enaydanov) [inline sdcm/cluster_baremetal.py:42]: We have a problem here: `self.parent_cluster` will be assigned later in `super().__init__()`, need to change it to just `parent_cluster`. I think it's came from my example, sorry about that :(

**Review by enaydanov — CHANGES_REQUESTED** [review]

**Review by enaydanov — DISMISSED** [review]

**Comment by k0machi** [comment]
@fruch @soyacz Looks like this PR is ready to be merged.

File: test-cases/performance/perf-regression-throughput-baremetal-5gb.yaml
Comment (fruch) [reply test-cases/performance/perf-regression-throughput-baremetal-5gb.yaml:13]: So check, having unrelated configuration would cause confusion

File: vars/perfRegressionBareMetalSinglePipeline.groovy
Comment (fruch) [inline vars/perfRegressionBareMetalSinglePipeline.groovy:1]: What does the Single mean here ?

File: vars/perfRegressionBareMetalSinglePipeline.groovy
Comment (fruch) [inline vars/perfRegressionBareMetalSinglePipeline.groovy:158]: All the pipeline duplicated just for this configuration of the runner ?

I would suggest introducing an optional runner_backend option to SCT for that

**Review by fruch — COMMENTED** [review]
Mostly comment about wording and configuration 

And the copy of the pipeline, that we should try to minimize.

File: vars/perfRegressionBareMetalSinglePipeline.groovy
Comment (k0machi) [reply vars/perfRegressionBareMetalSinglePipeline.groovy:1]: The previous pipeline was called Matrix, so this is the opposite of that.

File: vars/perfRegressionBareMetalSinglePipeline.groovy
Comment (k0machi) [reply vars/perfRegressionBareMetalSinglePipeline.groovy:158]: Yes, that would be a good thing to implement, should I do it with this PR or move it to a separate task?

File: vars/perfRegressionBareMetalSinglePipeline.groovy
Comment (fruch) [reply vars/perfRegressionBareMetalSinglePipeline.groovy:1]: You can just remove that word

File: vars/perfRegressionBareMetalSinglePipeline.groovy
Comment (fruch) [reply vars/perfRegressionBareMetalSinglePipeline.groovy:158]: It depends if it's a 20-30min work, let's do it now, if you estimate it to be more then that, let's write a separate task for it

File: vars/perfRegressionBareMetalSinglePipeline.groovy
Comment (k0machi) [reply vars/perfRegressionBareMetalSinglePipeline.groovy]: Done

File: vars/perfRegressionBareMetalSinglePipeline.groovy
Comment (k0machi) [reply vars/perfRegressionBareMetalSinglePipeline.groovy]: Reverted that, changed to adjusting `getCloudProviderFromBackend` to map `baremetal` to `aws`

Comment (k0machi) [reply vars/perfRegressionBareMetalSinglePipeline.groovy]: Seems to be working fine now, agent is properly selected and sct runner is created using old params

File: vars/perfRegressionBareMetalPipeline.groovy
Comment (fruch) [inline vars/perfRegressionBareMetalPipeline.groovy:2]: @k0machi 

one last thing, can you change code to run with `perfRegressionParallelPipeline.groovy`, it's now should have everything you were missing, i.e. creating sct runner

**Review by fruch — CHANGES_REQUESTED** [review]
I think now perfRegressionBareMetalPipeline.groovy is redundant, it doesn't have anything different then all other pipelines

File: vars/perfRegressionBareMetalPipeline.groovy
Comment (k0machi) [reply vars/perfRegressionBareMetalPipeline.groovy:2]: The parallel pipeline will try to deploy extra instances for read, write and mixed tests, where as this pipeline only does specified subtest in parameters. Since we only have one set of baremetal instances I think that's not going to work.

Comment (fruch) [reply vars/perfRegressionBareMetalPipeline.groovy:2]: it's depend on the list of subtest you supply to it, some tests have only one, and should work in you case.

I rather not duplicate pipelines, unless there's a real reason, so far I'm no convinced there is

**Comment by k0machi** [comment]
> I think now perfRegressionBareMetalPipeline.groovy is redundant, it doesn't have anything different then all other pipelines

That is indeed the case, I've removed the file and reapplied the original pipeline which seems to be [working](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/alexey/job/perf-regression-throughput-baremetal-example-test/31/)

**Review by fruch — APPROVED** [review]
LGTM

**Pull request merged**
